### PR TITLE
base.py / type hints

### PR DESCRIPTION
**Pull request by technicalgrp89**
tuna/miopen/driver/base.py - type annotations added.
mypy check cmd is added in to utils.groovy

### CODE REVIEW

File: tuna/miopen/driver/base.py [hunk 1]
```diff
@@ -54,12 +61,40 @@ def __init__(self, line=None, db_obj=None):
...
+
+  def parse_row(self, db_obj):
+    """Overloaded method.Defined in conv&bn driver child class"""
+
+  def set_cmd(self, data_type):
+    """Overloaded method.Defined in conv&bn driver child class"""
+
+  def config_set_defaults(self):
+    """Overloaded method.Defined in conv&bn driver child class"""
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py:76]:
```suggestion
raise NotImplementedError("Not implemented")
```

File: tuna/miopen/driver/base.py [hunk 2]
```diff
@@ -54,12 +61,40 @@ def __init__(self, line=None, db_obj=None):
       raise ValueError(
           "Error creating Driver. MIOpen Driver cmd line or db_obj required")
 
+  def parse_fdb_key(self, line):
+    """Overloaded method.Defined in conv&bn driver child class"""
+
+  def parse_row(self, db_obj):
+    """Overloaded method.Defined in conv&bn driver child class"""
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py:68]:
```suggestion
raise NotImplementedError("Not implemented")
```

File: tuna/miopen/driver/base.py [hunk 3]
```diff
@@ -54,12 +61,40 @@ def __init__(self, line=None, db_obj=None):
...
+  def set_cmd(self, data_type):
+    """Overloaded method.Defined in conv&bn driver child class"""
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py:72]:
```suggestion
raise NotImplementedError("Not implemented")
```

File: tuna/miopen/driver/base.py [hunk 4]
```diff
@@ -54,12 +61,40 @@ def __init__(self, line=None, db_obj=None):
...
+    """Overloaded method.Defined in conv&br driver child class"""
+
+  @staticmethod
+  def get_params(tok1):
+    """Overloaded method.Defined in conv&br driver child class"""
+
+  @staticmethod
+  def get_check_valid(tok1, tok2):
+    """Overloaded method.Defined in conv&br driver child class"""
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py:96]: raise NotImplementedError("Not implemented")

File: tuna/miopen/driver/base.py [hunk 5]
```diff
@@ -54,12 +61,40 @@ def __init__(self, line=None, db_obj=None):
...
   @staticmethod
-  def get_common_cols():
+  def test_skip_arg(tok1):
+    """Overloaded method.Defined in conv&br driver child class"""
+
+  @staticmethod
+  def get_params(tok1):
+    """Overloaded method.Defined in conv&br driver child class"""
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py:91]: raise NotImplementedError("Not implemented")

File: tuna/miopen/driver/base.py [hunk 6]
```diff
@@ -54,12 +61,40 @@ def __init__(self, line=None, db_obj=None):
...
+  @abstractmethod
+  def compose_weight_t(self):
+    """Overloaded method.Defined in conv&br driver child class"""
+
   @staticmethod
-  def get_common_cols():
+  def test_skip_arg(tok1):
+    """Overloaded method.Defined in conv&br driver child class"""
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py:86]: raise NotImplementedError("Not implemented")

File: tuna/miopen/driver/batchnorm.py [hunk 7]
```diff
@@ -79,6 +79,9 @@ def cmd(self, value):
            Supported cmds are: {SUPPORTED_BN_CMDS}')
     self._cmd = value
 
+  def compose_weight_t(self):
+    """ Overridden Method """
+
```
Comment (alexandraBara) [inline tuna/miopen/driver/batchnorm.py:89]: raise NotImplementedError("Not implemented")

File: tuna/miopen/driver/base.py [hunk 8]
```diff
@@ -111,13 +148,14 @@ def get_tensor_id(session, tensor_dict):
 
     return ret_id
 
-  def insert_tensor(self, tensor_dict):
+  def insert_tensor(self, tensor_dict: dict) -> Optional[int]:
     """Insert new row into tensor table and return primary key"""
+    ret_id: Optional[int] = 0
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]:
```suggestion
    ret_id: Optional[int] = -1
```

File: tuna/miopen/driver/base.py [hunk 9]
```diff
@@ -127,19 +165,22 @@ def insert_tensor(self, tensor_dict):
...
         ret_id = self.get_tensor_id(session, tensor_dict)
         LOGGER.info("Get Tensor: %s", ret_id)
-
     return ret_id
 
-  def get_input_t_id(self):
+  def get_input_t_id(self) -> Optional[int]:
     """Build 1 row in tensor table based on layout from fds param
        Details are mapped in metadata LAYOUT"""
+    ret_id: Optional[int] = 0
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]:
```suggestion
    ret_id: Optional[int] = -1
```

File: tuna/miopen/driver/base.py [hunk 10]
```diff
@@ -180,56 +221,68 @@ def decompose_input_t(self, db_obj):
 
     return True
 
-  def get_weight_t_id(self):
+  def get_weight_t_id(self) -> Optional[int]:
     """Build 1 row in tensor table based on layout from fds param
      Details are mapped in metadata LAYOUT"""
-    w_dict = self.compose_weight_t()
+    ret_id: Optional[int] = 0
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]:
```suggestion
    ret_id: Optional[int] = -1
```

File: tuna/miopen/driver/base.py [hunk 11]
```diff
@@ -85,12 +119,15 @@ def construct_driver_from_db(self, db_obj):
     return True
 
   @staticmethod
-  def get_tensor_id(session, tensor_dict):
+  def get_tensor_id(session: Session, tensor_dict: dict) -> Optional[int]:
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]: Should not be optional return, this should be a required int return

File: tuna/miopen/driver/base.py [hunk 12]
```diff
@@ -111,13 +148,14 @@ def get_tensor_id(session, tensor_dict):
 
     return ret_id
 
-  def insert_tensor(self, tensor_dict):
+  def insert_tensor(self, tensor_dict: dict) -> Optional[int]:
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]: same not optional

File: tuna/miopen/driver/base.py [hunk 13]
```diff
@@ -127,19 +165,22 @@ def insert_tensor(self, tensor_dict):
         session.rollback()
         ret_id = self.get_tensor_id(session, tensor_dict)
         LOGGER.info("Get Tensor: %s", ret_id)
-
     return ret_id
 
-  def get_input_t_id(self):
+  def get_input_t_id(self) -> Optional[int]:
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]: also not optional

File: tuna/miopen/driver/base.py [hunk 14]
```diff
@@ -180,56 +221,68 @@ def decompose_input_t(self, db_obj):
 
     return True
 
-  def get_weight_t_id(self):
+  def get_weight_t_id(self) -> Optional[int]:
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]: not optional

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: tuna/miopen/driver/base.py [hunk 15]
```diff
@@ -109,7 +114,7 @@ def construct_driver(self, line: str) -> bool:
     self.config_set_defaults()
     return True
 
-  def construct_driver_from_db(self, db_obj: Any) -> bool:
+  def construct_driver_from_db(self, db_obj: ConvolutionConfig) -> bool:
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]: leave it to Any, cause it can be both Conv and BatchNorn

File: tuna/miopen/driver/base.py [hunk 16]
```diff
@@ -200,7 +205,7 @@ def compose_input_t(self) -> Dict[str, int]:
 
     return i_dict
 
-  def decompose_input_t(self, db_obj) -> bool:
+  def decompose_input_t(self, db_obj: ConvolutionConfig) -> bool:
```
Comment (alexandraBara) [inline tuna/miopen/driver/base.py]: do not impose subclass, leave as Any

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: tuna/miopen/driver/base.py (hunk 15 above)
Comment (technicalgrp89) [reply tuna/miopen/driver/base.py]: I double checked via logs, Union type with ConvolutionConfig or batchnormconfig type, works fine

File: tuna/miopen/driver/base.py (hunk 16 above)
Comment (technicalgrp89) [reply tuna/miopen/driver/base.py]: defining Union[BNConnfig , ConvolutionConfig] is fine. tests are passing.

Comment (cderb) [reply tuna/miopen/driver/base.py]: This issue is that base may service additional subclasses, of which the two that are currently implemented may not be all inclusive. Additionally it is not the role of the base class to record all of its subclasses.

Comment (technicalgrp89) [reply tuna/miopen/driver/base.py]: Ok, Rolled back to Any type

File: tuna/miopen/driver/base.py (hunk 15 above)
Comment (technicalgrp89) [reply tuna/miopen/driver/base.py]: rolled back to Any

File: tuna/miopen/driver/base.py [hunk 17]
```diff
@@ -180,57 +226,71 @@ def decompose_input_t(self, db_obj):
...
+        f_digi_v = int(flag_value)
+      else:
+        f_digi_v = flag_value
+
+      flag_sh_value = self.get_params(flag_type)
+      if self.get_check_valid(flag_sh_value[0], f_digi_v):
+        setattr(self, flag_sh_value[0], f_digi_v)
       else:
         raise ValueError(
             f'Invalid command line arg for {self.cmd}: {tok1[0]} - {tok2} line: {line}'
```
Comment (cderb) [inline tuna/miopen/driver/base.py]: Variables for this print were renamed.

File: tuna/miopen/driver/base.py [hunk 18]
```diff
@@ -180,57 +226,71 @@ def decompose_input_t(self, db_obj):
...
         )
-
     return True
 
-  def to_dict(self):
+  def to_dict(self) -> Dict[str, int]:
```
Comment (cderb) [inline tuna/miopen/driver/base.py]: Convolution object has string fields like layout and direction. Not strictly Dict[str, int]. Currently Dict[str, Union[str, int]]

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/miopen/driver/base.py (hunk 17 above)
Comment (technicalgrp89) [reply tuna/miopen/driver/base.py:274]: variables , tok1 and tok2 line are not renamed, so the print statement remains unchanged

**Comment by alexandraBara** [comment]
Please add the cmd to lint the driver folder:
```
find miopen/ -type f -name "*.py" | xargs pylint -f parseable --max-args=8 --ignore-imports=no --indent-string '  '
```
You can add this in the groovy file right after the current pylint command (leave that one intact)

File: tuna/miopen/driver/base.py [hunk 19]
```diff
@@ -180,57 +226,71 @@ def decompose_input_t(self, db_obj):
...
-
     return True
 
-  def to_dict(self):
+  def to_dict(self) -> Dict[str, Union[str, int]]:
     """Return class to dictionary"""
-    copy_dict = {}
+    copy_dict: Dict[str, Union[str, int]] = {}
+    key: str
+    value: int
```
Comment (cderb) [inline tuna/miopen/driver/base.py]: int or str

**Review by cderb — CHANGES_REQUESTED** [review]

**Review by alexandraBara — APPROVED** [review]
Please address issue from @cderb

**Review by cderb — APPROVED** [review]

File: tuna/miopen/driver/base.py (hunk 4 above)
Comment (JehandadKhan) [reply tuna/miopen/driver/base.py:96]: Good suggestion!

**Pull request merged**